rew

Silumgaaar opened the pull request.

Silumgaaar committed:
rew

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+                }
+                workWitchDictionary(dictionarySelected);
+            } catch (IOException e){
+                e.printStackTrace();
```
Зачем пользователю видеть стек вызовов? Почему вообще тут эта ошибка? Думаю этот эксепшен логичнее будет обработать на месте, сюда вернув какой-то результат (может кастомную нашу ошибку). Это принцип единой ответственности.

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+                viewList(documents.getFileNames());
+                System.out.println(ConsoleConstants.EXIT_INFO);
+                System.out.print(ConsoleConstants.DICTIONARY_SELECTION);
+                nameFile = ConsoleConstants.user.next();
```
почему scanner это user? Это обман.

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+        this.documents = documents;
+        this.dictionary = dictionary;
+    }
+    public void work(){
```
Зачем этот метод? Просто прослойка?

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+                System.out.print(ConsoleConstants.DICTIONARY_SELECTION);
+                nameFile = ConsoleConstants.user.next();
+
+                if(nameFile.equals(ConsoleConstants.EXIT)){
```
Почему бы не добавить это в while? С небольшим рефакторингом.

NikolayKlyatishev commented on src/main/view/CommandsMenu.java:
```diff
@@ -0,0 +1,35 @@
+                return commandsMenu;
+            }
+        }
+        return null;
```
Почему null то? Ситуация вроде как исключительная. Возврашать null вообще считаю неправильно.

NikolayKlyatishev commented on src/main/view/CommandsMenu.java:
```diff
@@ -0,0 +1,35 @@
+    SEARCH("Ser - string search", "Ser"),
+    BACK("Back - back to menu", "Back");
+
+    private String title;
```
Поля могут быть final

NikolayKlyatishev commented on src/main/documentswork/DocumentsManager.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface DocumentsManager {
+    List<String> getFileNames() throws IOException;
+    HashMap<String, String> dictionaryHashMap(String fileName) throws IOException;
```
Не стоит использовать в именах методов типы. Если тип возвращаемого значения поменяется, придется менять и имя.

NikolayKlyatishev commented on src/main/dictionarywork/DictionaryManager.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.HashMap;
+
+public interface DictionaryManager {
+    public String search(HashMap<String,String> dictionarySelected,String key);
```
Он же и так public

NikolayKlyatishev commented on src/main/dictionarywork/DictionaryManager.java:
```diff
@@ -0,0 +1,9 @@
+public interface DictionaryManager {
+    public String search(HashMap<String,String> dictionarySelected,String key);
+    public HashMap<String,String> add(HashMap<String,String> dictionarySelected , String newKey, String newMeaning, String[] info);
+    public HashMap<String,String> remove(HashMap<String,String> dictionarySelected, String strDel);
```
Вот давай посмотрим на этот класс. Каждый метод тут принимает в качестве аргумента словарь. Внутри с этим словарем что-то происходит и все. 1 - зачем тут интерфейс, если всегда внутри Map и там всегда будет реализация одна. 2 - может тут что-то не так? Тут по сути даже объект не нужен. Можно и статическими методами все сделать, только в java так не делают. Думаю нужно выделить зоны ответственности. Подумать над уровнями приложения. Контроллер - сервис - репозиторий. У тебя сейчас есть контроллер и сервисы.

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+    public void work(){
+        mainMenu();
+    }
+    private void mainMenu(){
```
Вроде как судя по наименованию метода - он должен показывать главное меню. По сути оно делает все. Вся логика приложения тут. Абсолютно нарушен принцип единой ответственности. Выдели одну обязанность для каждого метода и давай обсудим.

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+            }
+        }
+    }
+    private void addInDictionary(HashMap<String,String> dictionarySelected) throws IOException {
```
Прочитай про паттерн стратегия. Каждое действие должно быть вынесено в отдельную стратегию поведения.

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+    private void removeInDictionary(HashMap<String,String> dictionarySelected) throws IOException {
+        System.out.print(ConsoleConstants.WORD_DELETE);
+        String[] info = documents.getInfoDictionary(nameFile);
+        dictionarySelected = dictionary.remove(dictionarySelected, ConsoleConstants.user.next());
```
Зачем в методе нужно возвращать значение? В каких случаях ставим конкретный тип, а в каких тип void?

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,114 @@
+    }
+    private void viewList(List<String> list){
+        for(String str : list){
+            System.out.println(str);
```
Не оптимально. Стоит сформировать строку и вывести ее. В твоем же случае ты постоянно повторяешь ресурсоемкую операцию вывода вместо формирования одной строки и вывода ее за 1 раз.

NikolayKlyatishev reviewed (commented):
В целом стало лучше.
Делай комиты меньше и атомарными.
Перед началом давай обсудим. Все прочитай, подумай и подходи. (предватирельно ознакомься со стратегией).

Silumgaaar committed:
Создание объектов словаря

Добавил код создания объектов словаря. В основной код пока не добавлял изменения

Silumgaaar committed:
rework

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+            Commands command = new Start();
+            command.execute(directory);
+            userChoice = ConsoleConstants.userChoice.next();
+            if (userChoice.equalsIgnoreCase("exit")) {
```
Это команда. Стоит включить ее в перечень стратегий

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+
+        mainMenu:
+        while(!exitProgram) {
+            Commands command = new Start();
```
Думаю стоит подумать над едином месте использования команд

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+                command.execute(dictionary);
+                System.out.println(viewMenu());
+
+                dictionaryMenu:
```
Не стоит ипользовать метки, хоть такая  возможность и есть. Думаю это признак неправильной структуры.

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+                System.out.println(viewMenu());
+
+                dictionaryMenu:
+                while (!command.getName().equalsIgnoreCase("Back")){
```
стоит так же вывести это в команды.

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+        this.listCommand = listCommand;
+    }
+    public void start(){
+        String userChoice;
```
Старайся объявлять переменные там, где будешь их использовать. Это облегчает процесс понимания кода

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+        boolean exitProgram = false;
+
+        DictionaryManager directory = new Dictionary();
+        nameFiles = directory.getDictionaries();
```
Подумать на абстрактном слое, к которому относится эта логика. Может это реализация?

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+        List<String>  nameFiles;
+        boolean exitProgram = false;
+
+        DictionaryManager directory = new Dictionary();
```
Является ли этот словарь состоянием системы?

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+                dictionaryMenu:
+                while (!command.getName().equalsIgnoreCase("Back")){
+                    userChoice = ConsoleConstants.userChoice.next();
+                    for(Commands c : listCommand){
```
Удобно ли работать постоянным перебором для поиска команды? Какая сложность алгоритма поиска в листе? Какая структура может предоставить возможность поиск за константное время?

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+
+    }
+    private boolean checkFile(List<String> nameFiles, String userChoice){
+        for(String str : nameFiles){
```
Какие методы выполняют аналогичную функцию в рамках интерфейса листа?

NikolayKlyatishev commented on src/main/view/ConsoleApp.java:
```diff
@@ -0,0 +1,82 @@
+        }
+        return false;
+    }
+    private String viewMenu(){
```
Нарушен принцип единой ответственности. Почему то составление меню ведется отдельно от команд.

NikolayKlyatishev commented on src/main/view/commands/Add.java:
```diff
@@ -0,0 +1,37 @@
+import main.view.ConsoleConstants;
+
+public class Add implements Commands{
+    public static final String INFO = "Add - add line";
```
А точно эти поля являются состоянием стратегии? При этом возникает сложность понимания, какие вобще команды есть. Это не удобно читать, редактировать.

NikolayKlyatishev commented on src/main/dictionarywork/Dictionary.java:
```diff
@@ -0,0 +1,160 @@
+
+    public boolean add(String newKey, String newValue){
+        if(checkAdd(newKey,newValue)){
+            dictionary.put(newKey, newValue);
```
А вдруг при записи файла произойдет IOEx?

NikolayKlyatishev commented on src/main/dictionarywork/Dictionary.java:
```diff
@@ -0,0 +1,160 @@
+                Files.write(Paths.get(patch), (entry.getKey() + DASH + entry.getValue() + "\n").getBytes(), StandardOpenOption.APPEND);
+            }
+        }catch (IOException e){
+            System.out.println(ConsoleConstants.FILE_NOT_FOUND);
```
Почему не найден? Других вариантов нет? Можешь ли ты согласно своей бизнес логике решать что делать на этом уровне абстракции?

NikolayKlyatishev commented on src/main/dictionarywork/Dictionary.java:
```diff
@@ -0,0 +1,160 @@
+            Path path = file.toPath();
+            list.addAll(Files.readAllLines(path));
+        } catch (IOException e){
+            System.out.println(ERROR_DIRECTORY);
```
Почему ошибка именно такая? Можно ли на текущем уровне абстракции решить проблему?

NikolayKlyatishev commented on src/main/dictionarywork/Dictionary.java:
```diff
@@ -0,0 +1,160 @@
+            fileOverWrite();
+            return true;
+        }
+        return false;
```
Этого достаточно для ответа?

NikolayKlyatishev commented on src/main/dictionarywork/Dictionary.java:
```diff
@@ -0,0 +1,160 @@
+    private String patch;
+    private String rulesKey;
+    private String rulesValue;
+    private final HashMap<String,String> dictionary = new HashMap<>();
```
В типе поля не принято указывать тип реализации, поскольку такое написание приводит к жесткой привязке к этому типу.
При этом паразительно схожее наименования 2 этих полей класса. Заставляет думать, что это одно и тоже. По сути 2-е должно быть List<Map<String, String>> ??

NikolayKlyatishev commented on src/main/dictionarywork/Dictionary.java:
```diff
@@ -0,0 +1,160 @@
+    private static final String DASH = "-";
+
+    public Dictionary(){
+        infoDirectory();
```
Есть 2 конструктора. Почему? Зачем 2?

NikolayKlyatishev commented on src/main/dictionarywork/Dictionary.java:
```diff
@@ -0,0 +1,160 @@
+            System.out.println(ConsoleConstants.FILE_NOT_FOUND);
+        }
+    }
+    private void infoDirectory(){                           // чтение файла Library, запись информации об именах файлов в директории в List<String> dictionaries
```
Комментарии в коде не приветствуются. Их стоит оставлять, чтобы объяснить зачем что-то реализовано именно так. Просто описывать функционал не нужно. Описание будет понятным при хорошем нейминге и архитектуре.

NikolayKlyatishev commented on src/main/view/commands/Commands.java:
```diff
@@ -0,0 +1,9 @@
+
+import main.dictionarywork.DictionaryManager;
+
+public interface Commands {
```
Методы getName и getInfo возвращают данные явно общей  сущности. Думаю стоит их объединить. (подумать как)

NikolayKlyatishev commented on src/main/view/ConsoleConstants.java:
```diff
@@ -0,0 +1,21 @@
+    public static final String EXIT = "Exit";
+    public static final String COMMAND_NOT_FOUND = "Command entered not found ";
+    public static final String STRING_NOT_FOUND = "String not found in dictionary ";
+    public static final String WORD_SEARCH = "Enter a word: ";
```
Возможно подобные описания относятся к конкретной бизнес логике конкретной команды?

NikolayKlyatishev commented on src/main/view/ConsoleConstants.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.Scanner;
+
+public class ConsoleConstants {
```
Считаю не очень хорошей практику складывать описания отдельно от места использования. Когда количество перевалит за 100 сойдешь с ума разбираться что где есть а чего еще нет.

NikolayKlyatishev reviewed (commented):
Просьба вносить исправления в отдельных ветках и небольшими атомарными комитами.

NikolayKlyatishev reviewed (changes requested):
Ревью
